src/Interpreters/Context.cpp

```diff
@@ -2414,4 +2415,66 @@ StorageID Context::resolveStorageIDImpl(StorageID storage_id, StorageNamespace w
     return StorageID::createEmpty();
 }
 
+std::shared_ptr<UUIDs> Context::getUUIDs()
+{
+    auto lock = getLock();
+    if (!uuids)
+        uuids = std::make_shared<UUIDs>();
+
+    return uuids;
+}
+
+std::shared_ptr<UUIDs> Context::getIgnoredUUIDs()
+{
+    auto lock = getLock();
+    if (!ignored_uuids)
+        ignored_uuids = std::make_shared<UUIDs>();
+
+    return ignored_uuids;
+}
+
+UUIDs::UUIDs() = default;
+UUIDs::~UUIDs() = default;
+
+std::vector<UUID> UUIDs::add(const std::vector<UUID> & uuids)
+{
+    std::lock_guard lock(mutex);
+    std::vector<UUID> intersection;
+
+    /// First check any presence of uuids in a map, return duplicates back if any
+    for(const auto & uuid : uuids)
+    {
+        if (map.find(uuid) != map.end())
+        {
+            intersection.emplace_back(uuid);
+        }
+    }
+
+    if (intersection.empty())
+    {
+        for(const auto & uuid : uuids)
+        {
+            map.emplace(uuid, nullptr);
+        }
+    }
+    return intersection;
+}
+
+std::vector<UUID> UUIDs::getList()
+{
+    std::lock_guard lock(mutex);
+    std::vector<UUID> v;
+    for (const auto &i: map) {
+        v.push_back(i.first);
+    }
+
+    return v;
+}
+
+/// Returns map of uuids, is only used in MergeTreeDataSelectExecutor to find a uuid
+UUIDs::Map UUIDs::getMap()
+{
+    std::lock_guard lock(mutex);
+    return map;
```

Comment:
you are returning a mutable map here? is it ok?

Reply:
no, good catch